File: iOS/GrokMode/UI/Voice Assistant/Models/VoiceSessionState.swift
//
//  VoiceSessionState.swift
//  GrokMode
//
//  Created by Abdulaziz Albahar on 12/14/25.
//

import Foundation

enum VoiceSessionState: Equatable {
    case disconnected
    case connecting
    case connected
    case listening
    case grokSpeaking(itemId: String?)
    case error(String)

    var isConnected: Bool {
        switch self {
        case .connected, .listening, .grokSpeaking: return true
        default: return false
        }
    }

    var isConnecting: Bool {
        if case .connecting = self { return true }
        return false
    }

    var isListening: Bool {
        if case .listening = self { return true }
        return false
    }

    var isGrokSpeaking: Bool {
        if case .grokSpeaking = self { return true }
        return false
    }

    var canStartListening: Bool {
        if case .connected = self { return true }
        return false
    }
}

File: Android/GrokMode/app/src/main/java/com/allensu/grokmode/voice/VoiceSessionState.kt
package com.allensu.grokmode.voice

sealed class VoiceSessionState {
    object Disconnected : VoiceSessionState()
    object Connecting : VoiceSessionState()
    object Connected : VoiceSessionState()
    object Listening : VoiceSessionState()
    data class Speaking(val itemId: String? = null) : VoiceSessionState()
    data class Error(val message: String) : VoiceSessionState()

    val isConnected: Boolean
        get() = this is Connected || this is Listening || this is Speaking

    val isConnecting: Boolean
        get() = this is Connecting

    val isListening: Boolean
        get() = this is Listening
}
